### DISCUSSION TITLE
How did auth credentials get cached via CDN?

### DISCUSSION THREAD

**Discussion by balazsorban44** [vercel/next.js#14136] [Help] [answered]
Hi, I am running a Next.js app on Azure App Services, with Azure CDN in front of it.

Doing this after a discussion I had with @Timer 
![https://i.imgur.com/tUCfl8u_d.webp?maxwidth=640&shape=thumb&fidelity=medium](https://i.imgur.com/tUCfl8u_d.webp?maxwidth=640&shape=thumb&fidelity=medium) 
Using https://github.com/auth0/nextjs-auth0 as an auth solution.

It has some pages using getStaticPaths with `fallback: true` and getStaticProps, but have also some data that is clientside fetched because it needs authentication. 
After logging in, my user information somehow got cached, and even if someone else logs in, they get my credentials.!? It looks like my credentials were saved in the CDN somehow. I would like to provide more info if necessary, also I was unsure where I should report this.

Someone has a good idea/experienced something similar?

**Comment by Timer** [comment, marked as answer]
It sounds like the CDN is caching something it shouldn't. What are the response headers sent for the specific page?

**Reply by balazsorban44** [reply]
![https://i.imgur.com/fg7367D.png](https://i.imgur.com/fg7367D.png)

This is from the index page, no getStaticProps/Paths on that page.

**Reply by Timer** [reply]
What pages are "sharing login state"? Do you have pages using `getServerSideProps`? You cannot return caching headers from these pages.

**Reply by balazsorban44** [reply]
![https://i.imgur.com/sp3Ns8J.png](https://i.imgur.com/sp3Ns8J.png) If this might be relevant, origin type is "Web app" on Azure. Shoud this be "Storage static website"?

No pages use `getServerSideProps`. I use the following `useUser` hook to access login state in pages:

```ts

export async function fetchUser(cookie = "") {
  if (typeof window !== "undefined" && window.__user) {
    return window.__user
  }

  const res = await fetch("/api/me", { headers: { cookie } })

  if (!res.ok) {
    delete window.__user
    return null
  }

  const json = await res.json()
  if (typeof window !== "undefined") {
    window.__user = json
  }
  return json
}

export function useUser({ required } : {required?: boolean} = {}) {
  const [loading, setLoading] = useState(
    () => !(typeof window !== "undefined" && window.__user)
  )
  const [user, setUser] = useState(() => {
    if (typeof window === "undefined") {
      return null
    }

    return window.__user || null
  })

  const router = useRouter()

  useEffect(() => {
    if (!loading && user) {
      return
    }
    setLoading(true)
    let isMounted = true

    fetchUser().then(user => {
      // Only set the user if the component is still mounted
      if (isMounted) {
        // When the user is not logged in but login is required
        if (required && !user) {
          window.location.href = `/api/login?redirectTo=${router.asPath}&returnUrl=${router.asPath}`
          return
        }
        setUser(user)
        setLoading(false)
      }
    })

    return () => {
      isMounted = false
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [])

  return ([user, loading] as [User | null, boolean])
}

```

This very well be an outdated method, I copied it from https://github.com/auth0/nextjs-auth0/blob/master/examples/basic-example/lib/user.js a while back, it looks like it now uses `React.Context` instead of `window.__user`

**Reply by balazsorban44** [reply]
UPDATE: I deployed it to Vercel, everything worked out-of-the-box, so yeah... It must be something with the config on Azure.

**Reply by Timer** [reply]
Do you see caching on the API endpoint (on Azure)? Try "custom origin".

**Reply by balazsorban44** [reply]
You mean if the pages under /api are cached on Azure?

I called an /api/foo endpoint, that returns information about the logged in user. The headers:

Azure deploy: x-cache: TCP_HIT
Vercel deploy: x-vercel-cache: MISS

**Reply by balazsorban44** [reply]
> Do you see caching on the API endpoint (on Azure)? Try "custom origin".

Tried "custom origin", did not help 😢

**Reply by balazsorban44** [reply]
Trying to set a no cache rule on /api/ now:

![https://i.imgur.com/JuCfTar.png](https://i.imgur.com/JuCfTar.png) 

~will update on if this helps.~

I think this solved the issue! Is it something needed to be done, or this should already be handled by Next.js and is a bug 🐛, then?

**Reply by Timer** [reply]
This seems to be an Azure bug to assume everything is cacheable by default. The cache headers for API routes do not instruct the CDN to cache—there must be a default setting somewhere.

You can verify this by running the server locally and inspecting the `cache-control` header returned by an API route.

---

Maybe configure the "Set if missing" option to not cache.
